### PR TITLE
[HPO] Ray HPO Automm and General

### PR DESCRIPTION
**Pull request by yinweisu**
*Issue #, if available:*

*Description of changes:*

This is a PR that cherry-picked Automm and general changes of PR https://github.com/awslabs/autogluon/pull/1696 to avoid having a giant PR that affects multiple modules at once.

This PR happens to still add quite some lines of code compared to the original one. That's because most of changes are on Automm side and we recently had a black reformat, which results in quite a lot line changes to merge in as well. The general hpo changes should be around 1000 lines.

* Added HPO functionality for AutoMM using ray tune + ray lightning.
  * Supports `random, bayes` searchers
  * Supports `FIFO, ASHA` schedulers
  * PBT/PB2 will be supported once ray lightning supports PT 1.6. Currently there's a bug in PT 1.6 stopping PBT/PB2 to work correctly
  * Extracted `_fit()` so that model specific configs will be generated in it, and `fit()` will generate general info that are model independent
  * Loading of the model has been divided into two parts to load general info, and config specific info
  * Checkpoint averaging is only done to the best trial once HPO is finished.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by sxjscience** [comment]
The logic looks good to me. I will approve once the CI gets fixed.

**Comment by szha** [comment]
Job PR-1790-2 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1790/2/index.html

**Review by zhiqiangdon — APPROVED** [review]
Thanks for adding HPO support!

**Review by sxjscience — APPROVED** [review]
LGTM. I will try it out after it gets merged. This is a critical PR!

File: core/src/autogluon/core/hpo/__init__.py
Comment (yinweisu) [reply core/src/autogluon/core/hpo/__init__.py:9]: It shouldn't as `AutommRayTuneAdapter` does not have any class or static variable that's package dependent

File: core/src/autogluon/core/hpo/ray_hpo.py
Comment (yinweisu) [reply core/src/autogluon/core/hpo/ray_hpo.py:106]: `minimum_gpu_per_trial ` is only passed to gpu resource calculator

File: core/src/autogluon/core/hpo/ray_hpo.py
Comment (yinweisu) [reply core/src/autogluon/core/hpo/ray_hpo.py:36]: I'm not putting HPO in a class because there's no state that needs to be maintained. In ray tune, the `run` function is also not wrapped in a class. Maybe we can have a `constant.py` file to store these constants.

File: core/src/autogluon/core/hpo/ray_hpo.py
Comment (yinweisu) [reply core/src/autogluon/core/hpo/ray_hpo.py:165]: Well, this is what our current hpo logic works. We supports for example `hyperparameter_tune_kwargs='auto'` for example. I think this is still important for a good user experience.

File: core/src/autogluon/core/hpo/ray_hpo.py
Comment (yinweisu) [reply core/src/autogluon/core/hpo/ray_hpo.py:189]: The adapter cannot be `None`. The logic here is if `resource_per_trial` is not None, means user wants to provide their own `resource_per_trial`, then we don't calculate it ourselves. Otherwise, we do the calculation. However, if it's `AutommRayTuneAdapter`, we do calculation no matter if user provided `resource_per_trial` or not

File: core/src/autogluon/core/hpo/ray_hpo.py
Comment (yinweisu) [reply core/src/autogluon/core/hpo/ray_hpo.py:300]: Good idea. It is closed and I'll need to check if it's being shipped with the recent ray release or not.

File: core/src/autogluon/core/hpo/ray_hpo.py
Comment (yinweisu) [reply core/src/autogluon/core/hpo/ray_hpo.py:371]: So you only specify `minimum_gpu_per_trial` if your trial requires gpu and you provided gpu in `total_resource`. Otherwise, this parameter will do nothing

File: core/src/autogluon/core/hpo/resources_calculator.py
Comment (gradientsky) [inline core/src/autogluon/core/hpo/resources_calculator.py:25]: nit: potentially a bad name for property. `type` is a [built-in function](https://docs.python.org/3/library/stdtypes.html#type-objects).

File: core/src/autogluon/core/hpo/resources_calculator.py
Comment (gradientsky) [inline core/src/autogluon/core/hpo/resources_calculator.py:100]: ditto: no memory assessments?

File: core/src/autogluon/core/space.py
Comment (gradientsky) [inline core/src/autogluon/core/space.py:130]: Design concerns: `Space` should not be aware of all possible ways to convert it. Conversion strategy should be decoupled from the entity. See [adapter pattern](https://refactoring.guru/design-patterns/adapter/python/example).
I would expect that we will pass `Space` to Ray-enabled sub-code, where it would be wrapped into Ray-specific adapter, which knows how to translate `Space` into ray spaces.

File: text/src/autogluon/text/automm/constants.py
Comment (gradientsky) [inline text/src/autogluon/text/automm/constants.py:58]: Can we use `higher_is_better` instead of manually hard-coding? See to [torchmetrics](https://torchmetrics.readthedocs.io/en/stable/pages/overview.html#metrics-and-hyperparameter-optimization).

File: text/src/autogluon/text/automm/utils.py
Comment (gradientsky) [inline text/src/autogluon/text/automm/utils.py:146]: use metric's [`higher_is_better`](https://torchmetrics.readthedocs.io/en/stable/pages/overview.html#metrics-and-hyperparameter-optimization) instead.

File: text/src/autogluon/text/automm/predictor.py
Comment (gradientsky) [inline text/src/autogluon/text/automm/predictor.py:585]: this method is too long. Can we split it into smaller logical units of work so it is easier to read and maintain?

File: core/src/autogluon/core/hpo/ray_hpo.py
Comment (yinweisu) [reply core/src/autogluon/core/hpo/ray_hpo.py:165]: I think you are right. We can only keep `num_trials`.

File: core/src/autogluon/core/hpo/ray_hpo.py
Comment (yinweisu) [reply core/src/autogluon/core/hpo/ray_hpo.py:189]: Good point. I'll pass `resource_per_trial` to each adapter's `get_resources_per_trial()` and let them decide what to do. For example, `TabularAdapter` could simply do nothing. This way we don't need to change the code if there are more adapters in the future

File: core/src/autogluon/core/hpo/resources_calculator.py
Comment (yinweisu) [reply core/src/autogluon/core/hpo/resources_calculator.py:71]: This becomes tricky here. Normally if a job is running on gpu, the mem usage is not that high. I guess what we want is `model_estimate_gpu_memory_usage`, and gpu ram are actually split into chunks(if we have multiple gpus) and how to properly calculate it is unclear.

File: core/src/autogluon/core/hpo/resources_calculator.py
Comment (yinweisu) [reply core/src/autogluon/core/hpo/resources_calculator.py:100]: automm doesn't currently has memory estimation, but maybe it's still a good idea to have the logic here in case

File: core/src/autogluon/core/space.py
Comment (yinweisu) [reply core/src/autogluon/core/space.py:130]: Good point!

Comment (yinweisu) [reply core/src/autogluon/core/space.py:130]: I guess this is not a strict adapter pattern because we don't want to implement the adapter for ray tune search space interface. This is more like a converter, but I can make it to be a standalone object instead of embedded in `Space` so that it's decoupled

File: text/src/autogluon/text/automm/constants.py
Comment (yinweisu) [reply text/src/autogluon/text/automm/constants.py:58]: I think this is out of scope in this PR, and the change could result in other changes in AutoMM. We can discuss it and improve it in a separate PR

Comment (sxjscience) [reply text/src/autogluon/text/automm/constants.py:58]: Yeah, we can refine this part in a separate PR. We can actually mark it as a TODO.

File: text/src/autogluon/text/automm/predictor.py
Comment (yinweisu) [reply text/src/autogluon/text/automm/predictor.py:585]: I think you refer to `_hyperparameter_tune()`This method is long because of many function signatures, for example `run()` and `_top_k_avetage()`, moving other things in to function will only short it by a little. For better readability, I've added comment on each logical operation.

File: core/src/autogluon/core/hpo/ray_hpo.py
Comment (yinweisu) [reply core/src/autogluon/core/hpo/ray_hpo.py:294]: https://github.com/awslabs/autogluon/issues/1798

File: text/src/autogluon/text/automm/constants.py
Comment (yinweisu) [reply text/src/autogluon/text/automm/constants.py:58]: Created an issue to track: https://github.com/awslabs/autogluon/issues/1799

**Review by gradientsky — APPROVED** [review]

**Comment by szha** [comment]
Job PR-1790-5 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1790/5/index.html

**Pull request merged**
